File: html/old/backup/sacor2_2007/rooms.py
#!/usr/bin/env python
# SACO Round 2 2007
# Solution by Marco Gallotta

W = int(raw_input("Enter width: "))
H = int(raw_input("Enter height: "))
store = [[False] * (W + 2)] 
for row in xrange(H):
    store.append([False] + map(lambda x: x == '0', raw_input("Enter row: ")) + [False])
store.append([False] * (W + 2)) 

def fill(x, y, n): 
    size = 0 
    if store[x][y] == True:
        store[x][y] = n 
        size = 1 
        for shift in [(0, -1), (0, 1), (-1, 0), (1, 0)]:
            size += fill(x + shift[0], y + shift[1], n)
    return size

(n, smallest, largest) = (0, 1000000000, 0)
for x in xrange(1, H + 1): 
    for y in xrange(1, W + 1): 
        if store[x][y] == True:
            n += 1
            size = fill(x, y, n + 1)
            smallest = min(smallest, size)
            largest  = max(largest,  size)

print "Number of rooms: %d" % n 
print "Smallest room:   %d" % smallest
print "Largest room:    %d" % largest

File: html/old/backup/sacor2_2007/rooms.cpp
// SACO 2nd Round 2007
// Rooms solution by Keegan Carruthers-Smith

#include <iostream>
#include <algorithm>

using namespace std;

bool seen[22][22]; // True if inspected cell
bool grid[22][22]; // False if wall. Grid is surrounded by walls for sentinals
int delta[4][2] = { {1,0}, {0,1}, {-1,0}, {0,-1} };

int dfs(int x, int y) {
    if ( seen[x][y] || !grid[x][y] )
        return 0;

    seen[x][y] = true;
    int sum = 1;
    for (int i = 0; i < 4; i++)
        sum += dfs( x + delta[i][0], y + delta[i][1] );

    return sum;
}

int main() {
    int width, height;
    int small, big, nRooms;
    char buf [21];
    
    big = nRooms = 0;
    small = 20*20;

    cout << "Enter width:  ";
    cin >> width;
    cout << "Enter height: ";
    cin >>  height;

    for (int x = 1; x <= height; x++) {
        cout << "Enter row: ";
        cin >> buf;
        for (int y = 1; y <= width; y++)
            grid[x][y] = buf[y-1] != '1';
    }

    for (int x = 1; x <= height; x++) {
        for (int y = 1; y <= width; y++) {
            int size = dfs(x,y);

            if (size != 0) {
                small = min(small, size);
                big   = max(big,   size);
                nRooms++;
            }
        }
    }

    cout << "Number of rooms: " << nRooms << endl;
    cout << "Smallest room:   " << small << endl;
    cout << "Largest room:    " << big << endl;
    
    return 0;
}
